File: rcc/communications/client/tcp/openrgb/effects/gaming.py
import re
from framework.singleton import singleton
from rcc.communications.client.tcp.openrgb.client.utils import DeviceType, RGBColor
from rcc.communications.client.tcp.openrgb.effects.base.abstract_effect import AbstractEffect


@singleton
class GamingEffect(AbstractEffect):
    """Gaming effect"""

    MAIN_COLOR = RGBColor.fromHEX("#0000FF")
    COLOR_MATCHING: list[re.Pattern, RGBColor] = {
        re.compile(r"Key: [1|2|3|4]"): RGBColor.fromHEX("#FFFF00"),
        re.compile(r"Key: [W|A|S|D]"): RGBColor.fromHEX("#FF0000"),
        re.compile(r"Key: Media Volume [-+]|Key: Media Mute"): RGBColor.fromHEX("#00FFFF"),
        re.compile(r"Key: Fan|Key: ROG"): RGBColor.fromHEX("#FF4000"),
    }

    def __init__(self):
        super().__init__("Gaming")

    def apply_effect(self):
        while self._is_running:  # pylint: disable=too-many-nested-blocks
            for dev in self._devices:
                colors = [self.MAIN_COLOR] * len(dev.leds)
                offset = 0
                if dev.type == DeviceType.KEYBOARD:
                    for zone in dev.zones:
                        for idx, led in enumerate(zone.leds):
                            for pattern, color in self.COLOR_MATCHING.items():
                                if pattern.fullmatch(led.name):
                                    colors[offset + idx] = color
                                    break
                        offset += len(zone.leds)
                self._set_colors(dev, colors)

            self._sleep(1)


GAMING_EFFECT = GamingEffect()

File: RogControlCenter/src/clients/tcp/open_rgb/effects/gaming.cpp
#include "../../../../../include/clients/tcp/open_rgb/effects/gaming.hpp"

const Color GamingEffect::MAIN_COLOR = Color::fromRgb("#000040");

const std::vector<std::pair<std::regex, Color>> GamingEffect::COLOR_MATCHING = {
	{std::regex("Key: [1|2|3|4]"), Color::fromRgb("#FFFF00")},
	{std::regex("Key: [W|A|S|D]"), Color::fromRgb("#FF0000")},
	{std::regex("Key: Media Volume [-+]|Key: Media Mute"), Color::fromRgb("#00FFFF")},
	{std::regex("Key: Fan|Key: ROG"), Color::fromRgb("#FF4000")},
};

GamingEffect::GamingEffect(Client& client) : AbstractEffect(client, "Gaming") {
}

void GamingEffect::apply_effect(const DeviceList& devices) {
	for (auto& dev : devices) {
		std::vector<Color> colors(dev.leds.size(), MAIN_COLOR);
		size_t offset = 0;

		if (dev.type == DeviceType::Keyboard) {
			for (auto& led : dev.leds) {
				for (auto& pair : COLOR_MATCHING) {
					if (std::regex_match(led.name, pair.first)) {
						colors[offset + led.idx] = pair.second;
						break;
					}
				}
			}
		}

		_set_colors(dev, colors);
	}
	while (_is_running) {
		_sleep(1.0);
	}
}
